### PR TITLE
Enhance function lookup with optimized trie data structure

### PR DESCRIPTION
**Pull request by zoo868e**
Changes:

- Added a new struct trie_t in defs.h to represent the trie.
- Created functions insert_trie_t and search_trie_t to handle trie operations.

### CODE REVIEW

File: src/defs.h [hunk 1]
```diff
@@ -163,3 +163,8 @@ typedef struct {
     char alias[MAX_VAR_LEN];
     int value;
 } constant_t;
+
+typedef struct trie_t{
```
Comment (jserv) [inline src/defs.h:167]: Replace `typedef struct trie_t` with `typedef struct` (anonymous structure).

**Review by jserv — CHANGES_REQUESTED** [review]
Append a newline at the end of src/defs.h and run `clang-format` to ensure the consistent style.

**Comment by jserv** [comment]
You shall read [CONTRIBUTING.md](https://github.com/jserv/shecc/blob/master/CONTRIBUTING.md) carefully and enforce the preferable style.

File: src/defs.h [hunk 2]
```diff
@@ -163,3 +163,8 @@ typedef struct {
     char alias[MAX_VAR_LEN];
     int value;
 } constant_t;
+
+typedef struct trie_t {
```
Comment (jserv) [inline src/defs.h:167]: Rewrite as `typedef struct {`.

Comment (zoo868e) [reply src/defs.h]: Done

File: src/globals.c [hunk 3]
```diff
@@ -6,6 +6,9 @@ int blocks_idx = 0;
 func_t *FUNCS;
 int funcs_idx = 0;
 
+trie_t *FUNCS_TRIE;
```
Comment (jserv) [inline src/globals.c:9]: Rename to `FUNC_TRIES` to emphasize a collection of trie data structure associated with functions.
In addition, add some comments.

File: src/globals.c [hunk 4]
```diff
@@ -35,6 +38,59 @@ char *elf_symtab;
 char *elf_strtab;
 char *elf_section;
 
+
```
Comment (jserv) [inline src/globals.c:41]: Avoid unnecessary blank lines.

File: src/globals.c [hunk 5]
```diff
@@ -35,6 +38,59 @@ char *elf_symtab;
...
+
+/*
+ * Enhence the find_func functions through trie_t struct.
+ *
+ * insert_trie_t:
+ * Parameters:
```
Comment (jserv) [inline src/globals.c:47]: Check https://www.kernel.org/doc/html/v4.10/doc-guide/kernel-doc.html for writing [Sphinx](http://www.sphinx-doc.org/) friendly comments.

File: src/globals.c [hunk 6]
```diff
@@ -35,6 +38,59 @@ char *elf_symtab;
...
+ *     *trie: A pointer to the trie where the name will be inserted.
+ *     *name: The name to be inserted into the trie.
+ *     funcs_index: The index of the pointer to the func_t. The index is
+ * recorded in a 1-indexed format, so it may need to be shifted if necessary.
+ * Description:
+ *     Inserts a new element into the trie structure.
+ */
+void insert_trie_t(trie_t *trie, char *name, int funcs_index)
```
Comment (jserv) [inline src/globals.c:55]: The function name should be `insert_trie`.

File: src/globals.c [hunk 7]
```diff
@@ -35,6 +38,59 @@ char *elf_symtab;
...
+ * search_trie_t:
+ * Parameters:
+ *     *trie: A pointer to the trie where the name will be searched.
+ *     *name: The name to be searched.
+ * Description:
+ *     Searches if the name already exists in the trie.
+ * Return value:
+ *     The index of the pointer to the func_t.
+ */
+int search_trie_t(trie_t *trie, char *name)
```
Comment (jserv) [inline src/globals.c:83]: Ditto. Rename to `find_trie`.

File: src/globals.c [hunk 8]
```diff
@@ -87,12 +143,14 @@ char *find_alias(char alias[])
...
-    int i;
-
-    /* return existing if found */
-    for (i = 0; i < funcs_idx; i++)
-        if (!strcmp(FUNCS[i].return_def.var_name, name))
-            return &FUNCS[i];
+    int index = search_trie_t(FUNCS_TRIE, name);
+    if (index)
+        return &FUNCS[index - 1]; /* return existing if found */
+    else
```
Comment (jserv) [inline src/globals.c:149]: The `else` statement is unnecessary.

**Review by jserv — CHANGES_REQUESTED** [review]
Squash git commits and mention the benefits (and some measurements) with the proposed change.

File: src/defs.h [hunk 9]
```diff
@@ -163,3 +163,8 @@ typedef struct {
     char alias[MAX_VAR_LEN];
     int value;
 } constant_t;
+
+typedef struct {
+    int index;
+    int next[128];
+} trie_t;
```
Comment (jserv) [inline src/defs.h:170]: See https://medium.com/@alexey.inkin/how-to-force-newline-at-end-of-files-and-why-you-should-do-it-fdf76d1d090e

**Review by jserv — CHANGES_REQUESTED** [review]
Don't check in the generated plots. Instead, summarize in git commit messages.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch.

**Comment by jserv** [comment]
> Rebase the latest `master` branch.

@zoo868e, Don't use `git merge`. Use `git rebase` instead.

**Comment by zoo868e** [comment]
I see, I'll do it again

**Comment by zoo868e** [comment]
Hi @jserv, I apologize for the inconvenience. I need assistance with the git rebase. In my local repository, I have already rebased the master branch to upstream/master. However, I'm unable to push the changes.

* b536f0e (HEAD -> master, temp) Stop check in results
* 359eff8 CI: Enforce newline at end of files
* 4b9fbf3 Stop check in results
* 94ee595 CI: Enforce newline at end of files
* fd2b64e (upstream/master) Add preliminary contributor guide
| * 4469fa9 (origin/master, origin/HEAD) Stop check in results
| * 70d4ee2 Add preliminary contributor guide
|/
* 7f6a7e7 CI: Enforce newline at end of files

**Comment by jserv** [comment]
> Hi @jserv, I apologize for the inconvenience. I need assistance with the git rebase. In my local repository, I have already rebased the master branch to upstream/master. However, I'm unable to push the changes.

You can simply expose the changes via `git format-patch` and then rework as new commits.

**Comment by zoo868e** [comment]
Hi @jserv, it appears that the commits have been squashed. Have I resolved the issue?

**Review by jserv — CHANGES_REQUESTED** [review]
Do not use markdown syntax in git commit messages. Instead, it is preferable to use plain text.

**Comment by zoo868e** [comment]
I see, let me fix it

**Review by jserv — CHANGES_REQUESTED** [review]
Execute `git commit --amend --author=` to change the authorship.
Since you can not resolve the potential SegFault introduced with this patch, you shall append the corresponding FIXME comments.

**Pull request closed without merging**